tests/utils/test_scripts.py

```
@@ -72,59 +97,48 @@ def _unsupervised_multi_gpu_cfg(cfg):
     return cfg
 
 
-def test_get_data_module_num_gpus_0(cfg):
+def test_get_data_module_num_gpus_0(cfg, mocker):
     cfg = _supervised_multi_gpu_cfg(cfg)
     # when num_gpus is set to 0, i.e. from a deprecated config
     cfg.training.num_gpus = 0
-    data_module = get_data_module(cfg, Mock(spec=BaseTrackingDataset))
+    mock_data_module_init = mocker.patch.object(BaseDataModule, '__init__', return_value=None)
+    get_data_module(cfg, Mock(spec=BaseTrackingDataset))
+
+    get_data_module(cfg, Mock(spec=BaseTrackingDataset))
```

Can remove both `get_data_module` calls, they don't appear to be used.

Can you say more about the motivation to use this mocker pattern? Why not just create a data module?

---

One is duplicative, removed.

The get_data_module has the side effect of calling the BaseDataModule constructor, which we then test assertions about.  But we don't need the returned object for this, we use the mocker object to assert what it was called with.

Reason for mocker: Previous code was running the real constructor, with a mock dataset as an input argument. This led to errors once the constructor started calling the _setup method. Mock dataset returns None for all attributes. The specific failure was about dataset.__len__ when the datamodule tries to split the dataset by length.

What the mocker does: replace data module constructor with a fake that does nothing. It just intercepts the args and allows you to assert things about them. This is the ideal thing for a "unit test". It tests just the logic in scripts.py, and exercises none of the code in datamodule.